### PR TITLE
[Datasets] Change `read_tfrecords` output from Pandas to Arrow format

### PR DESCRIPTION
**Pull request by c21**
Signed-off-by: Cheng Su <scnju13@gmail.com>

## Why are these changes needed?

This is to change `read_tfrecords` output from Pandas to Arrow format. From benchmark https://github.com/ray-project/ray/pull/30389, found the `read_tfrecords` is signigicantly slower than `write_tfrecords`.

E.g., for 1G TFRecords files (integers), the `write_tfrecords` took around 4 minutes, but `read_tfrecords` took more than 20 minutes. From flame graph, identified the bottleneck is coming from (1).creating Pandas dataframe from each row, and (2).Pandas `size_bytes()` call when adding each Pandas block to output buffer.

Change output from Pandas to Arrow (this is also the same format we use for `read_images`) fix the issue. The `read_tfrecords` latency drop from 24 minutes to less than 4 minutes (similar to `write_tfrecords`).

Before this fix:

Screen Shot 2022-11-16 at 4 43 40 PM

Screen Shot 2022-11-16 at 4 44 01 PM

After this fix:

Screen Shot 2022-11-16 at 6 34 24 PM

Screen Shot 2022-11-16 at 6 34 51 PM

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by c21** [comment]
The test is covered by https://github.com/ray-project/ray/pull/30389 .

**Review by clarkzinzow — APPROVED** [review]
Nice! I knew that there would be a perf improvement, but 6x is a good bit more than I had expected, that's awesome.

FYI, we have an open issue tracking this change, I'll assign it to you. https://github.com/ray-project/ray/issues/30003

File: python/ray/data/datasource/tfrecords_datasource.py [hunk 1]
```diff
@@ -64,13 +64,10 @@ def _write_block(
 
 def _convert_example_to_dict(
     example: "tf.train.Example",
-) -> Dict[str, Union[bytes, List[bytes], float, List[float], int, List[int]]]:
+) -> Dict[str, Union[List[List[bytes]], List[List[float]], List[List[int]]]]:
     record = {}
     for feature_name, feature in example.features.feature.items():
-        value = _get_feature_value(feature)
-        if len(value) == 1:
-            value = value[0]
```
Comment (c21) [inline python/ray/data/datasource/tfrecords_datasource.py:72]: I don't know why previously we have this special handling for the list with single element. Is it okay to remove? cc @bveeramani and @clarkzinzow.

Comment (bveeramani) [reply python/ray/data/datasource/tfrecords_datasource.py:72]: There's no way to create scalars with TFRecords. If your features represent scalars (as is usually the case), you still have to use list like `FloatList`.

Let's say you're working with the Iris dataset. Your TFRecords would look like this:

```
feature={
            "sepal.height": tf.train.Feature(float_list=tf.train.FloatList(value=[3.5])),
            "sepal.width": tf.train.Feature(float_list=tf.train.FloatList(value=[3.2])),
            "species": tf.train.Feature(bytes_list=tf.train.BytesList(value=[b"iris"])),
        }
```

If we didn't special case, your dataset that looks like:

```
Dataset(..., schema={"sepal.height": object, "sepal.width": object, "species": object}
```

This makes it impossible to use preprocessors and difficult to write UDFs.

What you probably want (which is what you get with the special casing) is a dataset like this:

```
Dataset(..., schema={"sepal.height": float, "sepal.width": float, "species": object}
```

I understand that special casing is generally bad, but based on the examples I've seen online, it seems like single-elements list represent scalars in the vast majority of cases. I've never seen a use case where a single-element list actually represents a list.

If we remove the special casing, `read_tfrecords` might not work the way users expect it to.

**Review by jianoaix — APPROVED** [review]
Nice. Also the profile is so useful per debugging.

File: python/ray/data/datasource/tfrecords_datasource.py (hunk 1 above)
Comment (c21) [reply python/ray/data/datasource/tfrecords_datasource.py]: @bveeramani - thanks for context. It makes sense to me, changed to maintain existing behavior.

**Comment by jianoaix** [comment]
Can we add one line here regarding the dataset format from reading tfrecords? https://docs.ray.io/en/master/data/dataset-internals.html#block-data-formats

**Comment by c21** [comment]
> Can we add one line here regarding the dataset format from reading tfrecords? https://docs.ray.io/en/master/data/dataset-internals.html#block-data-formats

@jianoaix - sure, added.

File: doc/source/data/dataset-internals.rst [hunk 2]
```diff
@@ -139,6 +139,7 @@ Different ways of creating Datasets leads to a different starting internal forma
 * Reading tabular files (Parquet, CSV, JSON) creates Arrow blocks initially.
 * Converting from Pandas, Dask, Modin, and Mars creates Pandas blocks initially.
 * Reading NumPy files or converting from NumPy ndarrays creates Arrow blocks.
+* Reading TFRecord file creates Arrow blocks.
```
Comment (c21) [inline doc/source/data/dataset-internals.rst:142]: This is the documentation update per request from https://github.com/ray-project/ray/pull/30390#issuecomment-1319339129, @stephanie-wang could you help check on behalf of ray-docs? Thanks.

**Review by stephanie-wang — APPROVED** [review]
Approving docs change.

**Pull request merged**